paas_app_charmer/charm.py

```diff
@@ -79,27 +60,9 @@ def __init__(self, framework: ops.Framework, framework_name: str) -> None:
         self._framework_name = framework_name
 
         self._secret_storage = KeySecretStorage(charm=self, key=f"{framework_name}_secret_key")
-        self._database_requirers = make_database_requirers(self, self.app.name)
-
-        requires = self.framework.meta.requires
-        if "redis" in requires and requires["redis"].interface_name == "redis":
-            self._redis = RedisRequires(charm=self, relation_name="redis")
-            self.framework.observe(self.on.redis_relation_updated, self._on_redis_relation_updated)
-        else:
-            self._redis = None
-
-        if "s3" in requires and requires["s3"].interface_name == "s3":
-            self._s3 = S3Requirer(charm=self, relation_name="s3", bucket_name=self.app.name)
-            self.framework.observe(self._s3.on.credentials_changed, self._on_s3_credential_changed)
-            self.framework.observe(self._s3.on.credentials_gone, self._on_s3_credential_gone)
-        else:
-            self._s3 = None
-
-        if "saml" in requires and requires["saml"].interface_name == "saml":
-            self._saml = SamlRequires(self)
-            self.framework.observe(self._saml.on.saml_data_available, self._on_saml_data_available)
-        else:
-            self._saml = None
+
+        self.charm_integrations = Integrations(self)
+        self.charm_integrations.register(self._on_generic_integration_event)
```

Comment:
I don't think I like it, I propose we go with the if-else ladder for now and get advice from @tonyandrewmeyer from the ops team on how to reduce it

Reply:
We can't do this because it takes away the option of the charm developer customising the code. All the observe need to be here and all the functions need to be defined on the class so that an adventurous Flask developer can change how they work and still be able to use the machinery we have built

Reply:
Sure, we can go without the refactor (I have it in the other [PR](https://github.com/canonical/paas-app-charmer/pull/39)).

However, I do not see why that would limit a user, everything can be overridden as it was before...

Reply:
(talked outside here) the reason is that there is only one callback function for all of the integrations

Reply:
It sounds like this is resolved now, but let me know if you'd still like input on anything :)